graph2tac/tfgnn/tasks.py

```diff
@@ -1281,66 +1337,94 @@ def create_inference_model(self, tactic_expand_bound: int, graph_constants: Grap
         proofstate_tactic_mask = tf.where(tf.expand_dims(no_context_proofstates, axis=-1),
                                           tf.expand_dims(no_argument_tactics_mask, axis=0),
                                           tf.expand_dims(all_tactics_mask, axis=0))
-
+        # [batch_size, tactic_expand_bound], [batch_size, tactic_expand_bound] 
         top_k_indices, top_k_values = self._top_k_tactics(tactic_logits=tactic_logits,
                                                           tactic_mask=proofstate_tactic_mask & tactic_mask,
                                                           tactic_expand_bound=tactic_expand_bound)
-
+        # [tactic_expand_bound*batch_size]
         tactic = tf.reshape(tf.transpose(top_k_indices), shape=(tf.size(top_k_indices),))
 
+        # [batch_size, None(context), hdim]
+        local_context_hidden = self._local_context_hidden(
+            scalar_proofstate_graph,
+            hidden_graph,
+        )
         # [batch_size, None(context), hdim]
         global_embeddings = self.global_embeddings(
             scalar_proofstate_graph.context['global_context_ids'],
             inference=True
         )
+        
+        batch_size = tf.shape(top_k_indices)[0]
 
         repeat_scalar_graph = RepeatScalarGraph(num_repetitions=tactic_expand_bound)
-        scalar_proofstate_graph = repeat_scalar_graph(scalar_proofstate_graph)  # noqa [ PyCallingNonCallable ]
+        # dimension of component features: [tactic_expand_bound*batch, ...]
         hidden_graph = repeat_scalar_graph(hidden_graph)  # noqa [ PyCallingNonCallable ]
-
-        hidden_state_sequences = self._hidden_state_sequences(hidden_graph=hidden_graph, tactic=tactic)
         # [tactic_expand_bound*batch, None(args), hdim]
+        hidden_state_sequences = self._hidden_state_sequences(hidden_graph=hidden_graph, tactic=tactic)
+        # [batch*tactic_expand_bound, None(args), hdim]
+        hidden_state_sequences = self.transpose_ragged_first_dim(
+            hidden_state_sequences, 
+            outer_dim=tactic_expand_bound, 
+            inner_dim=batch_size
+        )
```

Comment:
This is a bit awkward.  We stack all the graphs for the array in order to compute the tactic hidden embeddings, but really, there isn't a need to do this stacking, especially since it puts the tactic dimension first.  I probably can rewrite `_hidden_state_sequences`, but it wasn't a major bottleneck, so I didn't get to it.

Reply:
I decided to address this directly.